src/persist-client/src/internal/state_diff.rs

```diff
@@ -0,0 +1,576 @@
+pub struct StateFieldDiff<K, V> {
+    pub key: K,
+    pub val: StateFieldValDiff<V>,
+}
+
+#[derive(Debug)]
+#[cfg_attr(any(test, debug_assertions), derive(Clone, PartialEq))]
+pub struct StateDiff<T> {
+    pub(crate) applier_version: semver::Version,
+    pub(crate) seqno_from: SeqNo,
+    pub(crate) seqno_to: SeqNo,
+    pub(crate) latest_rollup_key: PartialRollupKey,
+    pub(crate) rollups: Vec<StateFieldDiff<SeqNo, PartialRollupKey>>,
+    pub(crate) last_gc_req: Vec<StateFieldDiff<(), SeqNo>>,
+    pub(crate) readers: Vec<StateFieldDiff<ReaderId, ReaderState<T>>>,
+    pub(crate) writers: Vec<StateFieldDiff<WriterId, WriterState>>,
+    pub(crate) since: Vec<StateFieldDiff<(), Antichain<T>>>,
+    pub(crate) spine: Vec<StateFieldDiff<HollowBatch<T>, ()>>,
+}
+
+impl<T: Timestamp + Lattice + Codec64> StateDiff<T> {
+    pub fn new(
+        applier_version: semver::Version,
+        seqno_from: SeqNo,
+        seqno_to: SeqNo,
+        latest_rollup_key: PartialRollupKey,
+    ) -> Self {
+        StateDiff {
+            applier_version,
+            seqno_from,
+            seqno_to,
+            latest_rollup_key,
+            rollups: Vec::default(),
+            last_gc_req: Vec::default(),
+            readers: Vec::default(),
+            writers: Vec::default(),
+            since: Vec::default(),
+            spine: Vec::default(),
+        }
+    }
+
+    pub fn from_diff<K, V, D>(from: &State<K, V, T, D>, to: &State<K, V, T, D>) -> Self {
+        // Deconstruct from and to so we get a compile failure if new
+        // fields are added.
+        let State {
+            applier_version: _,
+            shard_id: from_shard_id,
+            seqno: from_seqno,
+            collections:
+                StateCollections {
+                    last_gc_req: from_last_gc_req,
+                    rollups: from_rollups,
+                    readers: from_readers,
+                    writers: from_writers,
+                    trace: from_trace,
+                },
+            _phantom: _,
+        } = from;
+        let State {
+            applier_version: to_applier_version,
+            shard_id: to_shard_id,
+            seqno: to_seqno,
+            collections:
+                StateCollections {
+                    last_gc_req: to_last_gc_req,
+                    rollups: to_rollups,
+                    readers: to_readers,
+                    writers: to_writers,
+                    trace: to_trace,
+                },
+            _phantom: _,
+        } = to;
+        assert_eq!(from_shard_id, to_shard_id);
+
+        let (_, latest_rollup_key) = to.latest_rollup();
+        let mut diffs = Self::new(
+            to_applier_version.clone(),
+            *from_seqno,
+            *to_seqno,
+            latest_rollup_key.clone(),
+        );
+        diff_field_single(from_last_gc_req, to_last_gc_req, &mut diffs.last_gc_req);
+        diff_field_sorted_iter(from_rollups.iter(), to_rollups, &mut diffs.rollups);
```

Comment:
nit: I'm somewhat nervous because we require these to be sorted and rely on the `BTreeMap` for that. The `validate_roundtrip` that we do is probably a reasonable safeguard, but maybe we want debug assertions also in `diff_field_sorted_iter()`. Could also be overkill, of course.

Reply:
the debug assertion ended up super gross, but done!

```
                    let f_next = from.next();
                    debug_assert!(f_next.as_ref().map_or(true, |(fk_next, _)| fk_next > &fk));
                    f = f_next;
                    let t_next = to.next();
                    debug_assert!(t_next.as_ref().map_or(true, |(tk_next, _)| tk_next > &tk));
                    t = t_next;
```